### PR TITLE
Update memory snapshot harness documentation

### PR DESCRIPTION
**Pull request by danpoe**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- n/a Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: bd94f0ad8f8b7909406b242d4303c5e699a3b21b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117318447

File: doc/cprover-manual/goto-harness.md [hunk 1]
```diff
@@ -314,87 +314,121 @@ VERIFICATION SUCCESSFUL
...
+In order to take the snapshot with `memory-analyzer`, we need to first compile
+the program which `goto-gcc`, which produces a binary containing both native
+machine code and the corresponding goto program:
+
+```sh
+$ goto-gcc -g -o main.gb main.c
+```
+
+Then we can execute the program with `memory-analyzer` and take a snapshot at
+the specified breakpoint. The variables to include in the snapshot need to be
```
Comment (xbauch) [inline doc/cprover-manual/goto-harness.md]: .. variables to be included in ..

File: doc/cprover-manual/memory-analyzer.md [hunk 2]
```diff
@@ -2,40 +2,36 @@
...
+Calling `goto-gcc` instead of simply compiling with `gcc` or `goto-cc` produces
+an ELF binary with a goto section that contains the goto model (goto program +
+symbol table) [goto-cc-variants](../goto-cc/variants/).
 
-The memory analyzer supports two workflows to initiate the GDB with user code:
-either to run the code from a core-file or up to a break-point. If the user
-already has a core-file, they can specify it with the option `--core-file cf`.
-If the user knows the point of their program from where they want to run the
-analysis, they can specify it with the option `--breakpoint bp`. Only one of
+The memory analyzer supports two ways of running GDB on the user code: either t
```
Comment (xbauch) [inline doc/cprover-manual/memory-analyzer.md]: ...either _to_ run the code..

**Review by xbauch — APPROVED** [review]
Looks good.

File: doc/cprover-manual/goto-harness.md [hunk 3]
```diff
@@ -314,87 +314,121 @@ VERIFICATION SUCCESSFUL
...
-within a function. In that case we do not want the harness to instrument the
-whole function but rather to allow starting the execution from a specific
-initial location (specified via `--initial-location func[:<n>]`). Note that the
-initial location does not have to be the first instruction of a function: we can
-also specify the _location number_ `n` to set the initial location inside our
-function. The _location numbers_ do not have to coincide with the lines of the
-program code. To find the _location number_ run CBMC with
-`--show-goto-functions`. Most commonly, the _location number_ is the instruction
-of the break-point used to extract the program state for the memory snapshot.
+The `function-call` harness is used in situations in which we want to analyse a
```
Comment (peterschrammel) [inline doc/cprover-manual/goto-harness.md]:
```suggestion
The `function-call` harness is used in situations in which we want to analyze a
```

File: doc/cprover-manual/goto-harness.md [hunk 4]
```diff
@@ -314,87 +314,121 @@ VERIFICATION SUCCESSFUL
...
+function in an arbitrary environment. If we want to analyse a function starting
```
Comment (peterschrammel) [inline doc/cprover-manual/goto-harness.md]:
```suggestion
function in an arbitrary environment. If we want to analyze a function starting
```

File: doc/cprover-manual/goto-harness.md [hunk 5]
```diff
@@ -314,87 +314,121 @@ VERIFICATION SUCCESSFUL
...
+particular program location. The initial program location can be specified via
+the options `--initial-goto-location <function>[:<location-number>]` or
+`--initial-source-location <file>:<line-number>`.
 
 Say we want to check the assertion in the following code:
 
 ```C
 // main.c
 #include <assert.h>
+#include <stdlib.h>
```
Comment (peterschrammel) [inline doc/cprover-manual/goto-harness.md:333]: Have you actually run this code to check whether it works?

File: doc/cprover-manual/goto-harness.md [hunk 6]
```diff
@@ -314,87 +314,121 @@ VERIFICATION SUCCESSFUL
...
+  --symtab-snapshot \
+  --json-ui \
+  main.gb \
+  > snapshot.json
 ```
+
+We then generate a harness with `goto-harness` that behaves as if execution
+started from the state given by the memory snapshot at the specified program
+location. We further overapproximate the value returned by `get_random_value()`
+by havoc-ing the variable `x`.
```
Comment (peterschrammel) [inline doc/cprover-manual/goto-harness.md]:
```suggestion
by havocking the variable `x`.
```

File: doc/cprover-manual/memory-analyzer.md [hunk 7]
```diff
@@ -2,40 +2,36 @@
...
+Such a snapshot could be useful on its own: to check the values of variables at
+a particular program location. Furthermore, since the snapshot is a state of a
+valid concrete execution, it can also be used for subsequent analyses.
 
 ## Usage
 
-We assume that the user wants to inspect a binary executable compiled with
-debugging symbols and a symbol table information understandable by CBMC, e.g.
-(having `goto-gcc` on the `PATH`):
+In order to analyse a program with `memory-analyzer` it needs to be compiled
```
Comment (peterschrammel) [inline doc/cprover-manual/memory-analyzer.md]:
```suggestion
In order to analyze a program with `memory-analyzer` it needs to be compiled
```

**Review by peterschrammel — APPROVED** [review]

File: doc/cprover-manual/goto-harness.md (hunk 5 above)
Comment (danpoe) [reply doc/cprover-manual/goto-harness.md:333]: Yes

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b179f786a5666436820706cd7c4743ecab38d684).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117649125

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 65cca29be65a9f7fdd91df48e935a68c7f0d0454).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117646145
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Pull request merged**